### PR TITLE
Group L: poller hardening

### PR DESCRIPTION
**Pull request by jeremy**
## Summary

- **L1: Cursor monotonicity guard** -- `setActivitySince` and `setReadingsSince` now reject backwards-moving timestamps (clock skew) with a warning, preventing cursor regression
- **L2: Cursor save retry** -- `saveCursorsWithRetry` retries once after 1s on failure; logs error and continues if retry also fails
- **L3: Webhook concurrency limiter** -- Semaphore-based limit of 10 concurrent webhook dispatches with backpressure detection and queue overflow protection (100 max queued)
- **L4: Dispatch error classification** -- `onError` callback now logs structured metadata (agentId, recordingId, eventKind, senderId) and classifies errors as auth/network/routing/unknown
- **L5: State directory validation** -- Poller validates state directory is writable at startup, creating it if missing

## Test plan

- [x] `tests/poller-hardening.test.ts` -- cursor monotonicity (4 tests), save retry, state dir validation (2 tests)
- [x] `tests/webhooks.test.ts` -- Semaphore class (5 tests): acquire/release, FIFO ordering, pending count, rapid cycling
- [x] `tests/dispatch-enhanced.test.ts` -- enhanced onError logging with error classification (4 tests)
- [x] Existing `tests/dispatch.test.ts` and `tests/cursors.test.ts` pass unchanged
- [x] `npx tsc --noEmit` clean

### CODE REVIEW

File: src/inbound/cursors.ts [hunk 1]
```diff
@@ -67,16 +67,28 @@ export class CursorStore {
...
 
-  /** Update the activity feed cursor. */
+  /** Update the activity feed cursor. Only advances forward. */
   setActivitySince(since: string): void {
+    if (this.cursors.activitySince && since < this.cursors.activitySince) {
+      console.warn(
+        `[basecamp:cursors] clock skew detected: new activitySince (${since}) < existing (${this.cursors.activitySince}), skipping`,
+      );
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply src/inbound/cursors.ts]: Fixed — now uses `new Date(since) < new Date(this.cursors.activitySince)` for robust comparison regardless of timestamp format.

File: src/inbound/cursors.ts [hunk 2]
```diff
@@ -67,16 +67,28 @@ export class CursorStore {
...
 
-  /** Update the readings cursor. */
+  /** Update the readings cursor. Only advances forward. */
   setReadingsSince(since: string): void {
+    if (this.cursors.readingsSince && since < this.cursors.readingsSince) {
+      console.warn(
+        `[basecamp:cursors] clock skew detected: new readingsSince (${since}) < existing (${this.cursors.readingsSince}), skipping`,
+      );
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply src/inbound/cursors.ts]: Fixed — same Date-based comparison applied to setReadingsSince.

File: src/inbound/webhooks.ts [hunk 3]
```diff
@@ -18,6 +18,51 @@ import { dispatchBasecampEvent } from "../dispatch.js";
...
+  }
+
+  release(): void {
+    this.current--;
+    const next = this.queue.shift();
+    if (next) {
+      this.current++;
+      next();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply src/inbound/webhooks.ts:51]: Fixed — added guard that throws when release() is called with current <= 0, plus a test for this edge case.

File: tests/poller-hardening.test.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    const store = new CursorStore(tmpDir, "acct-1");
+    await store.load();
+    store.setActivitySince("2025-01-01T00:00:00Z");
+    await store.save();
+
+    const store2 = new CursorStore(tmpDir, "acct-1");
+    const loaded = await store2.load();
+    expect(loaded.activitySince).toBe("2025-01-01T00:00:00Z");
+  });
+});
```
[the remark this answers is not in the document]
Comment (jeremy) [reply tests/poller-hardening.test.ts:180]: Fixed — added two tests that exercise the retry pattern: one where save fails then succeeds on retry (verifies warn log, no error), and one where both attempts fail (verifies both warn and error logs, no throw).

File: tests/poller-hardening.test.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    const { constants } = await import("node:fs");
+
+    // tmpDir already exists
+    await mkdir(tmpDir, { recursive: true });
+    await access(tmpDir, constants.W_OK);
+
+    const s = await stat(tmpDir);
+    expect(s.isDirectory()).toBe(true);
+  });
+});
```
[the remark this answers is not in the document]
Comment (jeremy) [reply tests/poller-hardening.test.ts:222]: Added clarifying comment explaining these are unit tests of the fs operations the poller uses, not integration tests of the full poller startup.

File: tests/poller-hardening.test.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
+import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
+import { mkdtemp, rm, writeFile, chmod, stat } from "node:fs/promises";
```
[the remark this answers is not in the document]
Comment (jeremy) [reply tests/poller-hardening.test.ts]: Fixed — removed unused `writeFile` and `chmod` imports.

File: src/dispatch.ts [hunk 7]
```diff
@@ -143,6 +151,17 @@ export async function dispatchBasecampEvent(
...
 // ---------------------------------------------------------------------------
 
+/**
+ * Classify a dispatch error into a broad category for structured logging.
+ */
+function classifyDispatchError(err: unknown): string {
+  const s = String(err);
+  if (s.includes("401") || s.includes("403") || s.includes("Unauthorized") || s.includes("Forbidden")) return "auth";
+  if (s.includes("ETIMEDOUT") || s.includes("ECONNREFUSED") || s.includes("ECONNRESET") || s.includes("timeout")) return "network";
+  if (s.includes("no route") || s.includes("not found")) return "routing";
```
[the remark this answers is not in the document]
Comment (jeremy) [reply src/dispatch.ts]: Fixed — classifyDispatchError now checks structured error properties (status, statusCode, code) before falling back to message heuristics. Uses `\bno route\b` regex to avoid matching HTTP 404s as routing errors.

File: tests/dispatch-enhanced.test.ts [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      mockRuntime.channel.reply.dispatchReplyWithBufferedBlockDispatcher.mock
+        .calls[0][0];
+    const onError = call.dispatcherOptions.onError;
+
+    onError(new Error("something unexpected happened"));
+
+    const errorMsg = log.error.mock.calls[0][0];
+    expect(errorMsg).toContain("type=unknown");
+  });
+});
```
[the remark this answers is not in the document]
Comment (jeremy) [reply tests/dispatch-enhanced.test.ts:226]: Fixed — added tests for 403, ECONNRESET, structured status=401, and HTTP 404 (verified as unknown).

File: tests/webhooks.test.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    await sem.acquire();
+    await sem.acquire();
+    await sem.acquire();
+    expect(sem.pending).toBe(0);
+
+    sem.release();
+    sem.release();
+    sem.release();
+  });
+});
```
[the remark this answers is not in the document]
Comment (jeremy) [reply tests/webhooks.test.ts:148]: Fixed — added test for release() without acquire().

File: tests/dispatch-enhanced.test.ts [hunk 10]
```diff
@@ -0,0 +1,192 @@
+import { describe, it, expect, vi, beforeEach } from "vitest";
+import { dispatchBasecampEvent } from "../src/dispatch.js";
+import type {
+  BasecampInboundMessage,
+  ResolvedBasecampAccount,
+} from "../src/types.js";
+import { getBasecampRuntime } from "../src/runtime.js";
+import { resolvePersonaAccountId, resolveBasecampAccount } from "../src/config.js";
```
[the remark this answers is not in the document]
Comment (jeremy) [reply tests/dispatch-enhanced.test.ts]: Fixed — removed unused resolveBasecampAccount import.
